src/prefect/utilities/templating.py

```diff
@@ -0,0 +1,126 @@
+        return False
+
+
+UNSET: Unset = Unset()
+
+
+T = TypeVar("T", str, int, float, bool, dict, list)
+
+
+def apply_values(template: T, values: Dict[str, Any]) -> Union[T, Unset]:
+    """
+    Replaces placeholders in a template with values from a supplied dictionary.
+
+    Will recursively replace placeholders in dictionaries and lists.
+
+    If a value has no placeholders, it will be returned unchanged.
+
+    If a template contains only a single placeholder, the placeholder will be
+    fully replaced with the value.
+
+    If a template contains text before or after a placeholder or there are
+    multiple placeholders, the placeholders will be replaced with the
+    corresponding variable values.
+
+    If a template contains a placeholder that is not in `values`, UNSET will
+    be returned to signify that no placeholder replacement occurred. If
+    `template` is a dictionary that contains a key with a value of UNSET,
+    the key will be removed in the return value.
+
+    Args:
+        template: template to discover and replace values in
+        values: The values to apply to placeholders in the template
+
+    Returns:
+        The template with the values applied
+    """
+    variable_capture_regex = re.compile(r"({{\s*(\w+)\s*}})")
+
+    if isinstance(template, (int, float, bool, Unset)):
+        return template
+    if isinstance(template, str):
+        used_values = variable_capture_regex.findall(template)
+        if not used_values:
+            # If there are no variables, we can just use the value
+            return template
+        elif len(used_values) == 1 and used_values[0][0] == template:
+            # If there is only one variable with no surrounding text,
+            # we can replace it. If there is no variable value, we
+            # return UNSET to indicate that the value should not be included.
+            return values.get(used_values[0][1], UNSET)
+        else:
+            for full_match, variable_name in used_values:
+                if variable_name in values and values[variable_name] is not None:
+                    template = template.replace(
+                        full_match, values.get(variable_name, "")
+                    )
+            return template
+    elif isinstance(template, dict):
+        updated_template = {}
+        for key, value in template.items():
+            updated_value = apply_values(value, values)
+            if updated_value is not UNSET:
+                updated_template[key] = updated_value
+
+        return updated_template
+    elif isinstance(template, list):
+        updated_list = []
+        for value in template:
+            updated_value = apply_values(value, values)
+            if updated_value is not UNSET:
+                updated_list.append(updated_value)
+        return updated_list
+    else:
+        raise ValueError(f"Unexpected type: {type(template)}")
+
+
+@inject_client
+async def resolve_block_document_references(
```

Comment:
i might need some voice over on this one, i don't get why the `$ref` syntax is appearing

Reply:
The UI saves references to block documents in this structure:
```json
{
    "$ref": {
         "block_document_id":  <block_document_id>
    }
}
```

For block documents referenced from other block documents, the API will resolve the references so that the complete data is resolved and the client doesn't need to handle reference resolution. This structure isn't very user-friendly, so I expect that we'll add support for a `{{ prefect.blocks.<block_type_name>.<block_document_name> }}` syntax as well.

I've updated the docstring of this function to add more details. I plan on adding additional syntax support in another PR, but let me know if you think it should be included in this PR.

Reply:
ok gotcha; we can deal with the more natural spelling of block documents later